File: Application/src/main/java/com/shareyourproxy/api/rx/RxQuery.java
package com.shareyourproxy.api.rx;


import android.content.Context;
import android.support.annotation.NonNull;

import com.shareyourproxy.api.RestClient;
import com.shareyourproxy.api.domain.factory.UserFactory;
import com.shareyourproxy.api.domain.model.Channel;
import com.shareyourproxy.api.domain.model.Group;
import com.shareyourproxy.api.domain.model.GroupToggle;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.domain.realm.RealmUser;
import com.shareyourproxy.api.rx.command.eventcallback.GroupContactsUpdatedEventCallback;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;

import io.realm.Realm;
import io.realm.RealmResults;
import rx.Observable;
import rx.functions.Func1;

import static com.shareyourproxy.api.domain.factory.UserFactory.createModelUser;
import static com.shareyourproxy.api.rx.RxHelper.filterNullObject;

/**
 * Query the realm DB for data.
 */
public class RxQuery {

    public static HashMap<String, User> queryUserContacts(
        Context context, final HashSet<String> contactIds) {
        return Observable.just(context).map(new Func1<Context, HashMap<String, User>>() {
            @Override
            public HashMap<String, User> call(Context context) {
                HashMap<String, User> contacts;
                if (contactIds != null) {
                    contacts = new HashMap<>(contactIds.size());
                } else {
                    return new HashMap<>();
                }
                Realm realm = Realm.getInstance(context);
                realm.refresh();
                for (String contactId : contactIds) {
                    RealmUser realmUser =
                        realm.where(RealmUser.class).equalTo("id", contactId).findFirst();
                    if (realmUser != null) {
                        contacts.put(contactId, createModelUser(realmUser));
                    }
                }
                realm.close();
                return contacts;
            }
        }).toBlocking().first();
    }

    public static rx.Observable<HashMap<String, Channel>> queryPermissionedChannels(
        final User user, final String loggedInUserId) {
        return Observable.just(user)
            .map(getPermissionedChannels(loggedInUserId))
            .compose(RxHelper.<HashMap<String, Channel>>applySchedulers());
    }

    private static Func1<User, HashMap<String, Channel>> getPermissionedChannels(
        final String loggedInUserId) {
        return new Func1<User, HashMap<String, Channel>>() {
            @Override
            public HashMap<String, Channel> call(User contact) {
                ArrayList<String> permissionedIds = new ArrayList<>();
                HashMap<String, Channel> permissionedChannels = new HashMap<>();
                HashMap<String, Channel> channels = contact.channels();
                //escape early
                if (channels == null || channels.size() == 0) {
                    return permissionedChannels;
                } else {
                    HashMap<String, Group> groups = contact.groups();
                    //check the contacts groups for the logged in user and gather the channel
                    // Id's of that group
                    if (groups != null && groups.size() > 0) {
                        for (Group group : groups.values()) {
                            HashSet<String> contacts = group.contacts();
                            if (contacts != null) {
                                for (String contactId : contacts) {
                                    if (contactId.equals(loggedInUserId)) {
                                        permissionedIds.addAll(group.channels());
                                    }
                                }
                            }
                        }
                        // for the above key set data, find the channels associated
                        for (String channelId : permissionedIds) {
                            Channel channel = contact.channels().get(channelId);
                            if (channel != null) {
                                permissionedChannels.put(channel.id(), channel);
                            }
                        }
                    }
                    // add public channels
                    for (Channel channel : channels.values()) {
                        if (channel.isPublic()) {
                            permissionedChannels.put(channel.id(), channel);
                        }
                    }
                    return permissionedChannels;
                }
            }
        };
    }

    public static User queryUser(Context context, final String userId) {
        return Observable.just(context).map(getRealmUser(userId)).compose(RxHelper
            .<User>applySchedulers()).toBlocking().single();
    }

    public static Observable<Integer> getUserContactScore(Context context, final String userId) {
        return RestClient.getHerokuUserervice(context).userFollowerCount(userId).compose(RxHelper
            .<Integer>applySchedulers());
    }

    private static Func1<Context, User> getRealmUser(final String userId) {
        return new Func1<Context, User>() {
            @Override
            public User call(Context context) {
                return getRealmUser(context, userId);
            }
        };
    }

    public static User getRealmUser(Context context, final String userId) {
        Realm realm = Realm.getInstance(context);
        realm.refresh();
        RealmUser realmUser =
            realm.where(RealmUser.class).contains("id", userId).findFirst();
        User user = createModelUser(realmUser);
        realm.close();
        return user;
    }

    public static Observable<HashMap<String, User>> searchMatchingUsers(
        final Context context, final String queryName, final String userId) {
        return Observable.concat(
            searchLocalMatchingUsers(context, queryName, userId),
            searchRemoteMatchingUsers(context, queryName));

    }

    private static Observable<HashMap<String, User>> searchLocalMatchingUsers(
        Context context, final String queryName, final String userId) {
        return Observable.just(queryName).map(searchLocalUserString(context, userId));
    }

    private static Observable<HashMap<String, User>> searchRemoteMatchingUsers(
        Context context, final String queryName) {
        return Observable.just(queryName).map(searchRemoteUserString(context));
    }

    private static Func1<String, HashMap<String, User>> searchLocalUserString(
        final Context context, final String userId) {
        return new Func1<String, HashMap<String, User>>() {
            @Override
            public HashMap<String, User> call(String username) {
                return matchLocalUsers(context, userId, username);
            }
        };
    }

    private static HashMap<String, User> matchLocalUsers(
        Context context, final String userId, CharSequence constraint) {
        RealmResults<RealmUser> realmUsers;
        Realm realm = Realm.getInstance(context);
        realm.refresh();
        realm.beginTransaction();
        realmUsers = realm.where(RealmUser.class)
            .notEqualTo("id", userId)
            .beginsWith("fullName", constraint.toString(), false)
            .findAll();
        HashMap<String, User> users = UserFactory.createModelUsers(realmUsers);
        realm.commitTransaction();
        realm.close();
        return users;
    }

    private static Func1<String, HashMap<String, User>> searchRemoteUserString(
        final Context context) {
        return new Func1<String, HashMap<String, User>>() {
            @Override
            public HashMap<String, User> call(String queryName) {
                return RestClient.getHerokuUserervice(context)
                    .searchUsers(queryName).map(arrayToUserHashMap()).toBlocking().single();
            }
        };
    }

    public static Func1<ArrayList<User>, HashMap<String, User>> arrayToUserHashMap() {
        return new Func1<ArrayList<User>, HashMap<String, User>>() {
            @Override
            public HashMap<String, User> call(ArrayList<User> users) {
                HashMap<String, User> userMap = new HashMap<>(users.size());
                for (User user : users) {
                    userMap.put(user.id(), user);
                }
                return userMap;
            }
        };
    }

    public static List<GroupToggle> queryContactGroups(
        @NonNull final User user, final User selectedContact) {
        return Observable.from(user.groups().entrySet())
            .map(mapContacts(selectedContact))
            .toList()
            .toBlocking().single();
    }

    private static Func1<Map.Entry<String, Group>, GroupToggle> mapContacts(
        final User selectedContact) {
        return new Func1<Map.Entry<String, Group>, GroupToggle>() {
            @Override
            public GroupToggle call(Map.Entry<String, Group> groupEntry) {
                Group group = groupEntry.getValue();
                String contactId = selectedContact.id();
                HashSet<String> contacts = group.contacts();
                if (contacts != null &&
                    group.contacts().contains(contactId)) {
                    return new GroupToggle(group, true);
                }
                return new GroupToggle(group, false);
            }
        };
    }

    public static GroupContactsUpdatedEventCallback queryContactGroups(
        User user,
        ArrayList<GroupToggle> groupToggle, final String contactId) {
        return Observable.from(groupToggle).map(filterSelectedGroups())
            .filter(filterNullObject())
            .toList()
            .map(packageGroupContacts(user, contactId))
            .toBlocking().single();
    }

    private static Func1<GroupToggle, Group> filterSelectedGroups() {
        return new Func1<GroupToggle, Group>() {
            @Override
            public Group call(GroupToggle editContact) {
                return editContact.isChecked() ? editContact.getGroup() : null;
            }
        };
    }

    private static Func1<List<Group>, GroupContactsUpdatedEventCallback>
    packageGroupContacts(final User user, final String contactId) {
        return new Func1<List<Group>, GroupContactsUpdatedEventCallback>() {
            @Override
            public GroupContactsUpdatedEventCallback call(List<Group> groups) {
                return new GroupContactsUpdatedEventCallback(user, contactId, groups);
            }
        };
    }

}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/RxQuery.kt
package com.shareyourproxy.api.rx


import android.content.Context
import com.shareyourproxy.api.RestClient.herokuUserService
import com.shareyourproxy.api.domain.factory.UserFactory
import com.shareyourproxy.api.domain.factory.UserFactory.createModelUser
import com.shareyourproxy.api.domain.model.Channel
import com.shareyourproxy.api.domain.model.Group
import com.shareyourproxy.api.domain.model.GroupToggle
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.domain.realm.RealmUser
import com.shareyourproxy.api.rx.RxHelper.observeMain
import com.shareyourproxy.api.rx.command.eventcallback.GroupContactsUpdatedEventCallback
import io.realm.Case.INSENSITIVE
import io.realm.Realm
import io.realm.RealmResults
import rx.Observable
import rx.functions.Func1
import java.util.*

/**
 * Query the realm DB for data.
 */
object RxQuery {
    fun queryUserContacts(
            context: Context, contactIds: HashSet<String>?): HashMap<String, User> {
        return Observable.just<Context>(context).map<HashMap<String, User>>(
                getUserContacts(contactIds)).toBlocking().single()
    }

    private fun getUserContacts(contactIds: HashSet<String>?): Func1<Context, HashMap<String, User>> {
        return Func1 { context ->
            val contacts: HashMap<String, User>
            if (contactIds != null) {
                contacts = HashMap<String, User>(contactIds.size)
            } else {
                return@Func1 HashMap()
            }
            val realm = Realm.getInstance(context)
            realm.refresh()
            for (contactId in contactIds) {
                val realmUser = realm.where<RealmUser>(RealmUser::class.java).equalTo("id", contactId).findFirst()
                if (realmUser != null) {
                    contacts.put(contactId, createModelUser(realmUser))
                }
            }
            realm.close()
            contacts
        }
    }

    fun queryPermissionedChannels(
            user: User, loggedInUserId: String): Observable<HashMap<String, Channel>> {
        return Observable.just<User>(user).map<HashMap<String, Channel>>(getPermissionedChannels(loggedInUserId)).compose<HashMap<String, Channel>>(observeMain<HashMap<String, Channel>>())
    }

    private fun getPermissionedChannels(
            loggedInUserId: String): Func1<User, HashMap<String, Channel>> {
        return Func1 { contact ->
            val permissionedIds = ArrayList<String>()
            val permissionedChannels = HashMap<String, Channel>()
            val channels = contact.channels
            //escape early
            if (channels.size == 0) {
                permissionedChannels
            } else {
                val groups = contact.groups
                //check the contacts groups for the logged in user and gather the channel
                // Id's of that group
                if (groups != null && groups.size > 0) {
                    for (group in groups.values) {
                        val contacts = group.contacts
                        if (contacts != null) {
                            for (contactId in contacts) {
                                if (contactId == loggedInUserId) {
                                    permissionedIds.addAll(group.channels)
                                }
                            }
                        }
                    }
                    // for the above key set data, find the channels associated
                    for (channelId in permissionedIds) {
                        val channel = contact.channels[channelId]
                        if (channel != null) {
                            permissionedChannels.put(channel.id, channel)
                        }
                    }
                }
                // add public channels
                for (channel in channels.values) {
                    if (channel.isPublic) {
                        permissionedChannels.put(channel.id, channel)
                    }
                }
                permissionedChannels
            }
        }
    }

    fun queryUser(context: Context, userId: String): User {
        return Observable.just<Context>(context).map<User>(getRealmUser(userId)).compose<User>(observeMain<User>()).toBlocking().single()
    }

    fun getUserContactScore(userId: String): Observable<Int> {
        return herokuUserService.userFollowerCount(userId).compose<Int>(observeMain<Int>())
    }

    private fun getRealmUser(userId: String): Func1<Context, User> {
        return Func1 { context -> getRealmUser(context, userId) }
    }

    fun getRealmUser(context: Context, userId: String): User {
        val realm = Realm.getInstance(context)
        realm.refresh()
        val realmUser = realm.where<RealmUser>(RealmUser::class.java).contains("id", userId).findFirst()
        val user = createModelUser(realmUser)
        realm.close()
        return user
    }

    fun searchMatchingUsers(context: Context, queryName: String, userId: String): Observable<HashMap<String, User>> {
        return Observable.concat<HashMap<String, User>>(
                searchLocalMatchingUsers(context, queryName, userId),
                searchRemoteMatchingUsers(queryName))

    }

    private fun searchLocalMatchingUsers(context: Context, queryName: String, userId: String): Observable<HashMap<String, User>> {
        return Observable.just<String>(queryName).map<HashMap<String, User>>(searchLocalUserString(context, userId))
    }

    private fun searchRemoteMatchingUsers(queryName: String): Observable<HashMap<String, User>> {
        return Observable.just<String>(queryName).map<HashMap<String, User>>(searchRemoteUserString())
    }

    private fun searchLocalUserString(context: Context, userId: String): Func1<String, HashMap<String, User>> {
        return Func1 { username -> matchLocalUsers(context, userId, username) }
    }

    private fun matchLocalUsers(context: Context, userId: String, constraint: CharSequence): HashMap<String, User> {
        val realmUsers: RealmResults<RealmUser>
        val realm = Realm.getInstance(context)
        realm.refresh()
        realm.beginTransaction()
        realmUsers = realm.where<RealmUser>(RealmUser::class.java)
                .notEqualTo("id", userId)
                .beginsWith("fullName", constraint.toString(), INSENSITIVE)
                .findAll()
        val users = UserFactory.createModelUsers(realmUsers)
        realm.commitTransaction()
        realm.close()
        return users
    }

    private fun searchRemoteUserString(): Func1<String, HashMap<String, User>> {
        return Func1 { queryName ->
            herokuUserService.searchUsers(queryName)
                    .map<HashMap<String, User>>(arrayToUserHashMap())
                    .toBlocking().single()
        }
    }

    fun arrayToUserHashMap(): Func1<ArrayList<User>, HashMap<String, User>> {
        return Func1 { users ->
            val userMap = HashMap<String, User>(users.size)
            for (user in users) {
                userMap.put(user.id, user)
            }
            userMap
        }
    }

    fun queryContactGroups(user: User, selectedContact: User?): List<GroupToggle> {
        return Observable.from<Group>(user.groups.values)
                .map<GroupToggle>(mapContacts(selectedContact))
                .toList().toBlocking().single()
    }

    private fun mapContacts(selectedContact: User?): Func1<Group, GroupToggle> {
        return Func1 { group ->
            val contactId = selectedContact.id
            if (group.contacts.contains(contactId)) {
                return@Func1 GroupToggle(group, true)
            }
            GroupToggle(group, false)
        }
    }

    fun queryContactGroups(user: User, groupToggle: ArrayList<GroupToggle>, contactId: String):
            GroupContactsUpdatedEventCallback {
        return Observable.from<GroupToggle>(groupToggle)
                .map(filterSelectedGroups())
                .filter(RxHelper.filterNullObject())
                .toList()
                .map(packageGroupContacts(user, contactId))
                .toBlocking().single()
    }

    private fun filterSelectedGroups(): Func1<GroupToggle, Group> {
        return Func1 { editContact ->
            if (editContact.isChecked) editContact.group else null
        }
    }

    private fun packageGroupContacts(user: User, contactId: String):
            Func1<List<Group>, GroupContactsUpdatedEventCallback> {
        return Func1 { groups ->
            GroupContactsUpdatedEventCallback(user, contactId, groups)
        }
    }
}
